File: benchmark/python/numeric.py
"""Pure Python implementation of numeric computation functions."""

from typing import List


def find_primes(n: int) -> List[int]:
    """Find all prime numbers up to n using Sieve of Eratosthenes.
    
    Args:
        n: Upper limit for prime search
        
    Returns:
        List of all prime numbers <= n
    """
    if n < 2:
        return []
    
    # Initialize sieve: True means potentially prime
    is_prime = [True] * (n + 1)
    is_prime[0] = is_prime[1] = False
    
    # Sieve of Eratosthenes
    for i in range(2, int(n**0.5) + 1):
        if is_prime[i]:
            # Mark all multiples of i as not prime
            for j in range(i * i, n + 1, i):
                is_prime[j] = False
    
    # Collect all primes
    return [i for i in range(n + 1) if is_prime[i]]


def matrix_multiply(a: List[List[float]], b: List[List[float]]) -> List[List[float]]:
    """Multiply two matrices.
    
    Args:
        a: First matrix (m x n)
        b: Second matrix (n x p)
        
    Returns:
        Result matrix (m x p)
    """
    if not a or not b or not a[0] or not b[0]:
        return [[]]
    
    m = len(a)
    n = len(a[0])
    p = len(b[0])
    
    # Verify dimensions are compatible
    if len(b) != n:
        raise ValueError(f"Incompatible matrix dimensions: {m}x{n} and {len(b)}x{p}")
    
    # Initialize result matrix with zeros
    result = [[0.0 for _ in range(p)] for _ in range(m)]
    
    # Perform matrix multiplication
    for i in range(m):
        for j in range(p):
            for k in range(n):
                result[i][j] += a[i][k] * b[k][j]
    
    return result

File: benchmark/fortran_ext/numeric.f90
! Fortran implementation of numeric computation functions
! Using modern Fortran 90/95 features for better performance

module numeric_functions
    implicit none
    
contains

    ! Find all prime numbers up to n using Sieve of Eratosthenes
    subroutine find_primes_impl(n, primes, count)
        implicit none
        integer, intent(in) :: n
        integer, intent(out) :: count
        integer, dimension(n), intent(out) :: primes
        
        logical, dimension(0:n) :: is_prime
        integer :: i, j, sqrt_n
        
        ! Initialize all numbers as potentially prime
        is_prime = .true.
        is_prime(0) = .false.
        is_prime(1) = .false.
        
        ! Sieve of Eratosthenes
        sqrt_n = int(sqrt(real(n))) + 1
        do i = 2, sqrt_n
            if (is_prime(i)) then
                ! Mark all multiples of i as not prime
                do j = i*i, n, i
                    is_prime(j) = .false.
                end do
            end if
        end do
        
        ! Collect all primes
        count = 0
        do i = 2, n
            if (is_prime(i)) then
                count = count + 1
                primes(count) = i
            end if
        end do
    end subroutine find_primes_impl

    ! Matrix multiplication: C = A * B
    subroutine matrix_multiply_impl(a, b, c, m, n, p)
        implicit none
        integer, intent(in) :: m, n, p
        real(8), dimension(m, n), intent(in) :: a
        real(8), dimension(n, p), intent(in) :: b
        real(8), dimension(m, p), intent(out) :: c
        
        integer :: i, j, k
        real(8) :: temp
        
        ! Initialize result matrix
        c = 0.0d0
        
        ! Perform matrix multiplication with loop optimization
        do j = 1, p
            do k = 1, n
                do i = 1, m
                    c(i, j) = c(i, j) + a(i, k) * b(k, j)
                end do
            end do
        end do
    end subroutine matrix_multiply_impl

end module numeric_functions